compiler/ast/types.nim

```diff
@@ -1398,3 +1400,123 @@ proc isObjLackingTypeField*(typ: PType): bool {.inline.} =
   ## is not marked as ``.pure`` (the ``sfPure`` flags is not present on it)
   result = (typ.kind == tyObject) and ((tfFinal in typ.flags) and
       (typ[0] == nil) or isPureObject(typ))
+
+proc trav(marker: var IntSet, g: ModuleGraph, t: PType, search: PType,
+          isInd: bool): bool
+
+proc check(marker: var IntSet, g: ModuleGraph, t: PType, search: PType,
```

Comment:
Yes, I like it. Though I think `travNode` should also be named `productReachable` -- it's doing the same thing as `trav`, just with a record type's body instead of a type.